### ISSUE TITLE
Feature request: ability to configure weave

### ISSUE THREAD

**Issue by caseeker** [kubernetes/kops#1171]
It would be nice if we can configure weave through `kops edit cluster` command.
For example
```
networking:
  weave:
    ipalloc-range: 10.2.0.0/16
```
to set custom weave CIDR range to avoid possible IP clashing when cluster is launched in existing VPC.

**Comment by kris-nova** [comment]
I have heard this before, we would have to change the API for weave, and pass the value through to the manifest

Maybe we should wait for the api machinery in place first before looking at another API change?

**Comment by caseeker** [comment]
I'm wondering if we can templatize the [weave daemonset yaml file](https://github.com/kubernetes/kops/blob/c3d166805f350bea7325981f6672db76a7a86877/upup/models/cloudup/resources/addons/networking.weave/v1.8.1.20161130.yaml).

and add the IPALLOC_RANGE environment variable in daemonset configuration.
```
      containers:
          env:
            - name: IPALLOC_RANGE
              value: {{ WeaveConfig.IPALLOC_RANGE }}
```
I got this configuration from [Weave documenatation](https://www.weave.works/docs/net/latest/kube-addon/#configuration-options)

**Comment by chrislovecnm** [comment]
This is a great idea... @caseeker are you open to contributing this functionality?

Per @kris-nova's note this needs to pend on API machinery. We are going through a BIG refactor, but the design work could start today.

**Comment by caseeker** [comment]
@chrislovecnm 
I would love to contribute if some guidance is provided :)
I did try updating the daemonset yaml file in source and tested it, but that didn't work.

**Comment by chrislovecnm** [comment]
So this is the first place I would point you to -> https://github.com/kubernetes/kops/issues/1194

Get the use cases on here, and then work on a high-level design.  With this feature, the design can be super short.  Like a paragraph.  If you look at the PR for weave you will see the work I did for weave.  That needs to be extended with a yaml replacement as well.  I think we are doing that for a couple of other yaml templates, but we will need to walk you through it.  In a nutshell the yaml files can use go templates for replacement.  Basically call a go function. We are going to need to expand the API for this a bit as well.  I would perfer a value in the yaml cluster spec, instead of a command line switch.

Questions?

**Comment by caseeker** [comment]
@chrislovecnm sounds good, I'll prepare the PR,
Quick question for the implementation, I found out this [weave yaml file ](https://github.com/kubernetes/kops/blob/c3d166805f350bea7325981f6672db76a7a86877/upup/models/cloudup/resources/addons/networking.weave/v1.8.1.20161130.yaml) is not the one that gets used in cluster.
The cluster state bucket addons/networking.weave still contains v1.8.0.20161116.yaml even when I built the source with v1.8.1.20161130.yaml

Can you tell me if this is correct behavior? If so I'm wondering where v1.8.0.20161116.yaml is coming from as I could not locate it in source code.

**Comment by chrislovecnm** [comment]
It is the correct file :)  We have a bug in the versioning, which may be impacting you. Are you running on master?

**Comment by chrislovecnm** [comment]
https://github.com/kubernetes/kops/issues/1081 <- no design needed.  We need to rename the file v1.8.1.yaml

**Comment by caseeker** [comment]
I could make it working with updating the weave yaml file.
The problem was not from kops, I was updating on wrong src directory 😮

**Comment by chrislovecnm** [comment]
Well you figured it out!!!!

**Comment by justinsb** [comment]
I think this should use the `nonMasqueradeCIDR`, right?

**Comment by justinsb** [comment]
Is there a reason why weave isn't using the `nonMasqueradeCIDR` value @chrislovecnm ?

**Comment by chrislovecnm** [comment]
@justinsb don't have enough understanding to answer that.  Should we be??

**Comment by justinsb** [comment]
I think so.  Non-masquerade CIDR is the name we decided on for the overlay network CIDR: https://github.com/kubernetes/kubernetes/blob/master/cluster/aws/options.md

**Comment by justinsb** [comment]
I agree that we should support this configuration, and we may already have a suitable field in the form of nonMasqueradeCidr.  But, as per #1514, this isn't going to get fixed in 1.5.0, so I'm going to move this issue to 1.5.1.  (We will hopefully up the release cadence substantially once we get these big 1.5.0 features out).  Workarounds if you have a conflicting CIDR are to use kopeio-networking or calico.

And of course, contributions to fix this are welcome :-)

**Comment by jamesbucher** [comment]
Going to ask about this one. I went back and built a cluster with custom IPALLOC_RANGE for weave not to long ago and at the time did a custom build of kops to set it in the yaml. Now that there is already a template, classes and options I would like to tackle this.

However, I have a big question about the nonMasquerateCidr option. My understanding is that this parameter is passed to the kube-apiserver. However you can't actually just set the allocation range for weave to this because then it overlaps with the services IP address range and causes all sorts of issues. I know this from experience from trying it. https://github.com/weaveworks/weave/issues/2874#issuecomment-295354735 hints at this issue.

From a deployment I can see that kubeAPIServer.serviceClusterIPRange is 100.64.0.0/13 which doesn't necessarily conflict with the default IP address range of 100.64.0.0/10. Is the intent to derive the range automatically from the nonMasqueradeCidr range and make it distinct from the service IP address change?

I am currently running clusters with the IP address range for IPALLOC_RANGE completely separate from then nonMasqueradeCidr range and things are working fine. I am not really sure what implications of this are in detail. I just know it works. Does anyone know if there are implications for having weave run under a subnet that is entirely outside of the nonMasqueradeCidr range?

**Comment by jordanjennings** [comment]
@jamesbucher As mentioned in PR #2613 I also tried setting IPALLOC_RANGE to nonMasqueradeCidr and saw lots of issues due to the overlap in service IP range and the nonMasqueradeCidr. I'm surprised that it doesn't cause problems for other networking plugins to have the service IP range overlapping the pod IP range.

**Comment by austinmoore-** [comment]
@jamesbucher @jordanjennings I'm currently running Weave with `IPALLOC_RANGE` set to the pod's CIDR, not the full non-masquerade CIDR. AFAIK, Weave (and other networking solutions) should only come into play with the pod's IPs, since the Services are just iptable rules. I would expect problems to arise when Weave's CIDR overlaps with the Service IP range.

Someone correct me if I'm wrong, but I think that when setting `IPALLOC_RANGE` outside of nonMasqueradeCidr, your requests to other internal cluster resources may be masqueraded, and will have a different IP than the Pod's IP. This would violate some of the [Kubernetes Networking model requirements](https://kubernetes.io/docs/concepts/cluster-administration/networking/#kubernetes-model). (But that doesn't necessarily mean it won't work.)

**Comment by chrislovecnm** [comment]
@austinmoore- this is a problem that I introduced when I first implemented weave / CNI :)  I did not implement using the non-masquerade range with weave.

We need to figure out how to have a kubernetes upgrade work for users running the default `IPALLOC_RANGE`, and moving to the `nonMasqueradeCidr` range.  Upgrades are the one thing that is blocking this.  @jordanjennings tested upgrades a bunch and was not successful at getting weave to upgrade properly.

@bboreham what information can we provide you to give us more guidance.  We have weave running with the default IP range, and we upgrade the cluster and set `IPALLOC_RANGE` to 100.64.0.0/10 weave stop functioning.

@jordanjennings can you provide @bboreham more information?

**Comment by jordanjennings** [comment]
@chrislovecnm Not much info to provide, all networking was 100% broken when I tried upgrading. I didn't capture any logs or anything because I wasn't really sure where to look. I can certainly give it a shot again and try to capture more information.

But my main confusion here still remains around how it's OK for any networking provider to use the same range for pod IPs as it does for service IPs. What am I not understanding here?

@justinsb What is nonMasqueradeCIDR? I am very confused by the name, and why this field should map to pod IPs. I'm also confused why in the default kops configuration this overlaps with service IPs. Any additional info here would be great.

**Comment by chrislovecnm** [comment]
@jordanjennings the nonMasquereCIDR is the network range set aside for pod IP ranges.  I think the problem we are seeing is that the service range and the pod IP range are overlapping.  They cannot with CNI, and I am not certain why they are overlapping.

@bboreham / @caseydavenport are experts on these IP ranges with CNI if we could get him to comment that would be awesome.

In a nutshell, we seem to be having service IP ranges and pod IP ranges overlapping, which I think is really bad with CNI.  I think we have a bug with using the IP ranges that kops sets up with CNI.  If that is the case I am not certain how Calico is functioning as well.  I think weave is failing because it expects service and pod IP ranges to be separate.

**Comment by chrislovecnm** [comment]
Here are the values that we are setting up on a standard CNI install.  These values are part of the kops API, and typically get applied as flags to different components.

```yaml
  kubeAPIServer:
    serviceClusterIPRange: 100.64.0.0/13
  kubeControllerManager:
    # I do not think we need this with CNI
    clusterCIDR: 100.96.0.0/11
  kubeDNS:
    serverIP: 100.64.0.10
  kubeProxy:
    clusterCIDR: 100.96.0.0/11
  kubelet:
    clusterDNS: 100.64.0.10
    networkPluginName: cni
    nonMasqueradeCIDR: 100.64.0.0/10
  # these values are the base subnets that everything is setup from
  # I do not believe that these should overlap
  nonMasqueradeCIDR: 100.64.0.0/10
  serviceClusterIPRange: 100.64.0.0/13
```

**Comment by caseydavenport** [comment]
```
    # I do not think we need this with CNI
    clusterCIDR: 100.96.0.0/11
```

Not _strictly_ orthogonal with CNI - some CNI plugins / configuration still respect these allocations.

> the nonMasquereCIDR is the network range set aside for pod IP ranges. I think the problem we are seeing is that the service range and the pod IP range are overlapping. They cannot with CNI, and I am not certain why they are overlapping.

Yeah, can confirm - generally not a good idea to have overlapping Service / Pod / Node IP ranges.

**Comment by chrislovecnm** [comment]
@caseydavenport calico template is using `nonMasquereCIDR` directly, which can cause calico pod IP ranges to overlap with service IP ranges.

**Comment by caseydavenport** [comment]
Yeah, I think that should be considered a misconfiguration.  The default values definitely shouldn't overlap.  We might want some soft validation that warns if you've picked values that could cause problems (I'm not aware of any implementations where overlap is desired, but perhaps they're out there...)

**Comment by chrislovecnm** [comment]
So here is the word from @justinsb.  The pod IP range is:

```yaml
  kubeControllerManager:
    clusterCIDR: 100.96.0.0/11
```
All CNI providers should be using that CIDR for the pod IP range.  @jordanjennings is going to implement and test for weave.

**Comment by caseydavenport** [comment]
Ah yes, I think that is right.  I missed the distinction between nonMasqueradeCIDR and clusterCIDR.

The Calico manifests should be updated to use clusterCIDR :)

**Comment by chrislovecnm** [comment]
@jordanjennings here are some ideas on how we can approach this implementation:

1. Implement weave using `.KubeControllerManager.ClusterCIDR` in the weave manifest.  This code change will give you an easy way to do the difficult testing.
2. Implement a new API value `podClusterIPRange` with the proper validation.  The IP validation is going to require a review from someone who understands IP math.  If @justinsb / @bboreham or @caseydavenport could assist, that would be awesome.  I know my limitations, and IP math is something I have never figured out.
3. Migrate weave to use the `PodClusterIPRange`.

Thoughts anyone?  We could break up step 2 into API implementation and then validation, but I think it is simpler to implement together.

**Comment by jordanjennings** [comment]
@chrislovecnm 👍  I'm testing out step 1 right now

**Comment by chrislovecnm** [comment]
@ottoyiu, since you have been doing a lot of the calico work as of late, wanted to loop you in on this issue.  

TLDR - Calico should not be using the non-masq range, but should be using `.KubeControllerManager.ClusterCIDR`, since it overlaps with the service IP range. These range overlaps are NOT good.

cc: @blakebarnett

**Comment by ottoyiu** [comment]
@chrislovecnm thanks for the heads up. I'll test with the ClusterCIDR and report back. This ticket might be explaining these weird one-off service IP issues I've been experiencing.

**Comment by jamesbucher** [comment]
@austinmoore- I am not 100% sure that having a Weave Network outside of the nonMasqueradeCidr actually results in requests being MASQUERADED. I had 2 pods up and running in a cluster with NonMasqueradeCidr set to: 100.64.0.0/10 and the IPALLOC_RANGE for weavenet set to 192.168.240/21. With pods on separate nodes I am able to ping across pods and the requests appear to be coming directly through. I am able to use the Pod IP address (for the server) from a client (curl) to connect and the server (nginx). The server also actually sees the correct client podIP (verified from the nginx logs).

Correct me if I am wrong but wouldn't masqueraded traffic disallow this as it is a form of NAT (either the server or client would appear to have the wrong IP address)?

Regardless of weather or not this works its probably a really bad idea to have the internal weave network operate outside of the NonMasqueradeCidr range.

**Comment by chrislovecnm** [comment]
@jamesbucher we are not planning on having weave outside of the range. We need to run it in the pod range. We worked out which subnet is for services and which subnet is for pods, which are subsets of the non masq cidr.

Make sense?

Btw the first PR is in for weave. The other providers require updates and testing as well.

**Comment by jamesbucher** [comment]
@chrislovecnm Makes sense. I just wanted to note that it does actually work. I actually went in today and updated our clusters to ensure that Weave's CIDR corresponds to the podClusterIPRange so that we don't have to do some crazy network migration at a later date when our clusters are taking more traffic.
